### PR TITLE
[SPARK-30326][SQL] Raise exception if analyzer exceed max iterations

### PR DESCRIPTION
**Pull request by Eric5553**
### What changes were proposed in this pull request?
Enhance RuleExecutor strategy to take different actions when exceeding max iterations. And raise exception if analyzer exceed max iterations.

### Why are the changes needed?
Currently, both analyzer and optimizer just log warning message if rule execution exceed max iterations. They should have different behavior. Analyzer should raise exception to indicates the plan is not fixed after max iterations, while optimizer just log warning to keep the current plan. This is more feasible after SPARK-30138 was introduced.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Add test in AnalysisSuite

### CODE REVIEW

**Comment by Eric5553** [comment]
@cloud-fan @dongjoon-hyun @HyukjinKwon
Would you please help review, thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/AnalyzerRuleStrategySuite.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+package org.apache.spark.sql
+
+import org.apache.spark.{LocalSparkContext, SparkConf, SparkContext, SparkFunSuite}
+import org.apache.spark.sql.catalyst.errors.TreeNodeException
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.TestSparkSession
+
+class AnalyzerRuleStrategySuite extends SparkFunSuite with LocalSparkContext {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/AnalyzerRuleStrategySuite.scala]: Can you write tests in `AnalysisSuite` without a Spark session?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -169,7 +169,7 @@ class Analyzer(
 
   def resolver: Resolver = conf.resolver
 
-  protected val fixedPoint = FixedPoint(maxIterations)
+  protected val fixedPoint = FixedPoint(maxIterations, errorOnExceed = true)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: > Analyzer should raise exception to indicates logical plan resolve failed, while optimizer just log warning to keep the current plan. 

Does exceeding max iteration indicate analyzer fails logical plan resolving? Doesn't it just mean the plan is not fixed after max iterations? It sounds not equal.

Comment (Eric5553) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Yeah, I see. I've updated the descriptions of PR & JIRA. Thanks for pointed it out :-)

File: sql/core/src/test/scala/org/apache/spark/sql/AnalyzerRuleStrategySuite.scala (hunk 1 above)
Comment (Eric5553) [reply sql/core/src/test/scala/org/apache/spark/sql/AnalyzerRuleStrategySuite.scala]: Sure, I'll give a try. Thanks!

**Comment by cloud-fan** [comment]
If the analyzer hits max iteration, and the plan is resolved. shall we log warning or fail?

**Comment by Eric5553** [comment]
@cloud-fan Good catch!

IMO, literally the description `The max number of iterations the analyzer runs.` infers we should try to iterate the plan to the maxIteration inclusively with no error or warning. But the current code will raise exception or log warning at the max iteration even the plan is fixed. This apply to both analyzer & optimizer.

I tried to fix it in 015e97218b60da907e49ad9bebb208a276a13354, and also added a corresponding test case in RuleExecutorSuite. The case will fail(raise exception) with original code.

What's your opinion? Please correct me if I'm wrong. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/AnalyzerRuleStrategySuite.scala (hunk 1 above)
Comment (Eric5553) [reply sql/core/src/test/scala/org/apache/spark/sql/AnalyzerRuleStrategySuite.scala]: @maropu I moved the test to AnalysisSuite without using Spark session in     4f3ac3ac19bdfa18bd37fe950c1b6b7412406551.  Does this looks properly? Thanks!

**Comment by Eric5553** [comment]
@cloud-fan @maropu @viirya  Would you please help review the latest change? Thanks!

**Comment by maropu** [comment]
> Analyzer should raise exception to indicates the plan is not fixed after max iterations

I'm currently not sure that this feature is useful... If a logical plan has unresolved nodes for that case, I think Spark should throw an analysis exception in `CheckAnalysis`. Do you mean that is not enough?

**Comment by gatorsmile** [comment]
@maropu I think the exception threw by CheckAnalysis is confusing to end users. The unresolved plans are not caused by the user queries. Instead, they need to increase the value of SQLConf.ANALYZER_MAX_ITERATIONS.

**Comment by gatorsmile** [comment]
> If the analyzer hits max iteration, and the plan is resolved, shall we log warning or fail?

in this case, keep the existing behavior?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala [hunk 3]
```diff
@@ -45,16 +45,23 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
...
    * execution reaches fix point (i.e. converge) before maxIterations, it will stop.
    */
-  abstract class Strategy { def maxIterations: Int }
+  abstract class Strategy {
+
+    /** The maximum number of executions. */
+    def maxIterations: Int
+
+    /** Whether to throw exception when exceeding the maximum number. */
+    def errorOnExceed: Boolean
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: Why did you add this in not `RuleExecutor` but `Strategy`?

Comment (Eric5553) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: > Why did you add this in not `RuleExecutor` but `Strategy`?

@maropu The setting `SQLConf.ANALYZER_MAX_ITERATIONS` only controls a subset of Analyzer Rule batches (Hints, Substitution, Resolution and Cleanup). For other Rules with `Once` strategy, the original RuleExecutor logic will not raise exception or log warning. Which means within Analyzer, it will take different actions for different Batches/Strategy. So I think the `errorOnExceed` behavior is more like binding with maxIteration in Strategy. By doing this, we could have finer granularity to support the feature and provide better flexibility to control behaviors for different batches in the future. Thanks!

**Comment by Eric5553** [comment]
> > If the analyzer hits max iteration, and the plan is resolved, shall we log warning or fail?
> 
> in this case, keep the existing behavior?

Sure, I've reverted the code to keep existing behavior for this case. Thanks for review! @gatorsmile

**Comment by Eric5553** [comment]
@cloud-fan @maropu @viirya Would you please help take a look? Thanks!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala [hunk 4]
```diff
@@ -47,6 +47,19 @@ trait AnalysisTest extends PlanTest {
     }
   }
 
+  protected def makeAnalyzerWithConf(conf: SQLConf): Analyzer = {
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala]: We cannot test it with `getAnalyzer()` and `withSQLConf` in AnalysisSuite (we need this helper method?)?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -169,7 +169,7 @@ class Analyzer(
 
   def resolver: Resolver = conf.resolver
 
-  protected val fixedPoint = FixedPoint(maxIterations)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Can you leave some comments here about why we need this change?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala (hunk 4 above)
Comment (Eric5553) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala]: @maropu The strategy field `fixedPoint` of Analyzer/Optimizer is determined when analyzer/optimizer initialized, thus SQLConf will not affect their behavior during runtime. `getAnalyzer()` only support makeAnalyzer with CASE_SENSITIVE config, while I think adding more config such as ANALYZER_MAX_ITERATIONS to `getAnalyzer()` will be too specific. That’s why I introduced this helper method with SQLConf input parameter, which will also support future test case requirements like ANALYZER_MAX_ITERATIONS does. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 5 above)
Comment (Eric5553) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: @maropu Sure, added in commit 8a20a5c622d4b52980036c620d7e848925e43f38. Thanks for point it out!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala (hunk 3 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: Ur, I see. Thanks. Since this method is used only once now in `AnalysisSuite`, can you move this into `AnaysisSuite`?

Comment (Eric5553) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: > Ur, I see. Thanks. Since this method is used only once now in `AnalysisSuite`, can you move this into `AnaysisSuite`?

Sure, will do :-)

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala (hunk 4 above)
Comment (Eric5553) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisTest.scala]: @maropu I moved the helper method in commit ab82c4136144e3bd2b00ce94e87927f7e9eb93f9

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala (hunk 3 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: (Oh, I noticed now that I left the comment in the wrong position, sorry...)

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 6]
```diff
@@ -737,4 +756,23 @@ class AnalysisSuite extends AnalysisTest with Matchers {
       CollectMetrics("evt1", count :: Nil, tblB))
     assertAnalysisError(query, "Multiple definitions of observed metrics" :: "evt1" :: Nil)
   }
+
+  test("Analysis exceed max iterations") {
+    // RuleExecutor only throw exception or log warning when the rule is supposed to run
+    // more than once.
+    val maxIterations = 2
+    val conf = new SQLConf().copy(SQLConf.ANALYZER_MAX_ITERATIONS -> maxIterations)
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: I think its ok to inline the analyzer like this;
```
  test("Analysis exceed max iterations") {
    // RuleExecutor only throw exception or log warning when the rule is supposed to run
    // more than once.
    val maxIterations = 2
    val testAnalyzer = new Analyzer(
      new SessionCatalog(new InMemoryCatalog, FunctionRegistry.builtin, conf),
      new SQLConf().copy(SQLConf.ANALYZER_MAX_ITERATIONS -> maxIterations))

    val plan = testRelation2.select(
      $"a" / Literal(2) as "div1",
      $"a" / $"b" as "div2",
      $"a" / $"c" as "div3",
      $"a" / $"d" as "div4",
      $"e" / $"e" as "div5")

    val message = intercept[TreeNodeException[LogicalPlan]] {
      testAnalyzer.execute(plan)
    }.getMessage
    assert(message.startsWith(s"Max iterations ($maxIterations) reached for batch Resolution"))
  }
```

**Comment by maropu** [comment]
ok to test

**Comment by maropu** [comment]
Looks fine now. I'll leave this to the other more qualified reviewers. @cloud-fan @viirya

**Comment by Eric5553** [comment]
@maropu Thanks for the detailed review!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala (hunk 6 above)
Comment (Eric5553) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: Yeah, I see. It's more concise.

**Comment by Eric5553** [comment]
@cloud-fan @viirya @gatorsmile  Would you please revisit the changes, thanks!

**Comment by Eric5553** [comment]
@cloud-fan @viirya @gatorsmile Would you please revisit the changes, thanks!

**Comment by gatorsmile** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala [hunk 7]
```diff
@@ -156,7 +163,7 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
           // Only log if this is a rule that is supposed to run more than once.
           if (iteration != 2) {
             val message = s"Max iterations (${iteration - 1}) reached for batch ${batch.name}"
-            if (Utils.isTesting) {
+            if (Utils.isTesting || batch.strategy.errorOnExceed) {
               throw new TreeNodeException(curPlan, message, null)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: In the message, we need to explicitly tell users how to resolve this exception. Increasing the value of SQLConf.ANALYZER_MAX_ITERATIONS.key

**Comment by gatorsmile** [comment]
LGTM except a comment about the message

**Comment by gatorsmile** [comment]
cc @cloud-fan @maryannxue

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala (hunk 7 above)
Comment (Eric5553) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: IMO, here is the common logic in `RuleExecutor` to handle `errorOnExceed` strategy, which should not include specific knowledge `ANALYZER_MAX_ITERATIONS` of inheritance class `analyzer`.  This will lead to potential bug when we introduce more MAX_ITERATION setting?

Maybe we can use a common phrase here like 'increasing the value of corresponding SQLConf setting'. Or an alternative way of adding a string hint field in `Strategy`, and set it in `Analyzer` together with `errorOnExceed` flag?

What do you think ? @gatorsmile  Thanks a lot!

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: We need a clear message for end users. Usability is more important. 

Regarding the implementation, I think we should avoid over-engineering. Just make it simple. Adding the message postfix in abstract class Strategy looks fine to me.

Comment (Eric5553) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: I see, thanks for the clarification. :-)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala [hunk 8]
```diff
@@ -155,8 +162,9 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
         if (iteration > batch.strategy.maxIterations) {
           // Only log if this is a rule that is supposed to run more than once.
           if (iteration != 2) {
-            val message = s"Max iterations (${iteration - 1}) reached for batch ${batch.name}"
-            if (Utils.isTesting) {
+            val message = s"Max iterations (${iteration - 1}) reached for batch ${batch.name}, " +
+              s"increasing the value of '${SQLConf.ANALYZER_MAX_ITERATIONS.key}'."
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: Actually, this change is not right. The optimizer will also issue this message.

Comment (Eric5553) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: Opps, sorry.
I updated the logic in cdc48795b8ad39359186830a7cc0c5be3dc169aa. Extend abstract class `Strategy` to support setting different hint setting key in Analyzer or Optimizer.
Would you please take time to review? Thanks!

**Comment by gatorsmile** [comment]
add to whitelist

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala [hunk 9]
```diff
@@ -155,8 +168,14 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
         if (iteration > batch.strategy.maxIterations) {
           // Only log if this is a rule that is supposed to run more than once.
           if (iteration != 2) {
-            val message = s"Max iterations (${iteration - 1}) reached for batch ${batch.name}"
-            if (Utils.isTesting) {
+            val message = batch.strategy.maxIterationsSetting match {
+              case setting: String if setting != null =>
+                s"Max iterations (${iteration - 1}) reached for batch ${batch.name}, " +
+                  s"increasing the value of '${setting}'."
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: nit: `please set '${setting}' to a larget value`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala [hunk 10]
```diff
@@ -155,8 +168,14 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
         if (iteration > batch.strategy.maxIterations) {
           // Only log if this is a rule that is supposed to run more than once.
           if (iteration != 2) {
-            val message = s"Max iterations (${iteration - 1}) reached for batch ${batch.name}"
-            if (Utils.isTesting) {
+            val message = batch.strategy.maxIterationsSetting match {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala]: nit:
```
val endingMsg = if (batch.strategy.maxIterationsSetting == null) {
  "."
} else {
  ", please set '${setting}' to a larget value"
}
val message = s"Max iterations (${iteration - 1}) reached for batch ${batch.name}$endingMsg"
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 11]
```diff
@@ -745,4 +746,27 @@ class AnalysisSuite extends AnalysisTest with Matchers {
...
+    val testAnalyzer = new Analyzer(
+      new SessionCatalog(new InMemoryCatalog, FunctionRegistry.builtin, conf),
+      new SQLConf().copy(SQLConf.ANALYZER_MAX_ITERATIONS -> maxIterations))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: nit: just pass `conf`

**Comment by Eric5553** [comment]
@cloud-fan Thanks for the review! bff75bc6e2e877c08113062dc471458caa5cb996 was to address the comments.

**Comment by Eric5553** [comment]
retest this please

**Review by cloud-fan — APPROVED** [review]

**Comment by Eric5553** [comment]
retest this please

**Comment by cloud-fan** [comment]
This needs to be in 3.0. In Spark 2.4 the analyzer max iteration was controlled by `SQLConf.OPTIMIZER_MAX_ITERATIONS`. In 3.0, we add a new config for it, and it's possible that existing queries set `SQLConf.OPTIMIZER_MAX_ITERATIONS` to a large value and they fail analysis after upgrading to Spark 3.0.

With this PR, they can get a clear error message to set the new config.

Thanks, merging to master/3.0!

**Comment by Eric5553** [comment]
Thank you all !!
